staging/src/k8s.io/component-base/config/serializer/encode.go

```diff
@@ -0,0 +1,103 @@
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+)
+
+// EncodingFormat is a high-level, typed constant configuring how an object should be encoded.
+type EncodingFormat string
+
+const (
+	// ContentTypeYAML encodes a config file in the YAML format.
+	ContentTypeYAML EncodingFormat = "application/yaml"
+	// ContentTypeJSON encodes a config file in compact JSON format.
+	ContentTypeJSON EncodingFormat = "application/json"
+)
+
+// NewStrictYAMLJSONSerializer creates a new implementation of the StrictYAMLJSONSerializer interface.
+func NewStrictYAMLJSONSerializer(scheme *runtime.Scheme) StrictYAMLJSONSerializer {
+	knownEncoders := map[EncodingFormat]runtime.Encoder{}
+	codecs := serializer.NewCodecFactory(scheme)
+	for _, serializerInfo := range codecs.SupportedMediaTypes() {
+		switch serializerInfo.MediaType {
+		case "application/json":
+			knownEncoders[ContentTypeJSON] = serializerInfo.Serializer
+		case "application/yaml":
+			knownEncoders[ContentTypeYAML] = serializerInfo.Serializer
+		}
+	}
+
+	// Create the strict deserializer, and creates a decoder later.
+	strictYAMLSerializer := json.NewStrictYAMLSerializer(json.DefaultMetaFactory, scheme, scheme)
+	universalDecoder := codecs.CodecForVersions(nil, strictYAMLSerializer, nil, runtime.InternalGroupVersioner)
+
+	return &strictYAMLJSONSerializer{scheme, knownEncoders, universalDecoder}
+}
+
+// StrictYAMLJSONSerializer provides encoding/decoding for a config file with YAML or JSON format.
+type StrictYAMLJSONSerializer interface {
+	DecodeInto([]byte, runtime.Object) error
+	Encode(EncodingFormat, schema.GroupVersion, runtime.Object) ([]byte, error)
+}
+
+// strictYAMLJSONSerializer implements the StrictYAMLJSONSerializer interface.
+var _ StrictYAMLJSONSerializer = &strictYAMLJSONSerializer{}
+
+type strictYAMLJSONSerializer struct {
+	scheme        *runtime.Scheme
+	knownEncoders map[EncodingFormat]runtime.Encoder
+	decoder       runtime.Decoder
+}
+
+// DecodeInto decodes bytes into a pointer of the desired type.
+func (cs *strictYAMLJSONSerializer) DecodeInto(data []byte, into runtime.Object) error {
+	out, gvk, err := cs.decoder.Decode(data, nil, into)
+	// return a structured error if the group was registered with the scheme but the version was unrecognized.
+	if gvk != nil && err != nil {
+		if cs.scheme.IsGroupRegistered(gvk.Group) && !cs.scheme.IsVersionRegistered(gvk.GroupVersion()) {
+			return NewUnrecognizedVersionError(*gvk, data)
+		}
+	}
+	if err != nil {
+		return err
+	}
+	if out != into {
+		return fmt.Errorf("unable to decode %s into %T", gvk, reflect.TypeOf(into))
+	}
+	return nil
+}
+
+// Encode encodes the object into a byte slice for the specific format and version.
+func (cs *strictYAMLJSONSerializer) Encode(format EncodingFormat, gv schema.GroupVersion, obj runtime.Object) ([]byte, error) {
```

Comment:
there's nothing special/different about encoding objects strictly, right? it's unclear to me why we're coupling conversion and serialization into "strictYAMLJSONSerializer"

Reply:
yes. But in my opinion. we use `strictYAMLJSONSerializer` obj to transfer `Encode()`,  which can distingush it

Reply:
@luxas WDYT?

Reply:
I am really not sure why we have a new encoder api.  That’s a higher level “get me an encoder if the type”.  Which we already have in codec dactory